warehouse/query-core/src/main/java/datawave/query/attributes/UniqueGranularity.java

```diff
@@ -0,0 +1,113 @@
+package datawave.query.attributes;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import datawave.data.normalizer.DateNormalizer;
+import org.apache.log4j.Logger;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.function.Function;
+
+/**
+ * Represents different levels of granularity supported by the {@code #unique()} function. This class is also responsible for providing the functionality to
+ * transform values such that they conform to the specified granularity.
+ */
+public enum UniqueGranularity {
+    
+    /**
+     * A {@link UniqueGranularity} implementation that will always return the original value.
+     */
+    ALL("ALL", Function.identity()),
+    
+    /**
+     * A {@link UniqueGranularity} implementation that, if provided a datetime value, will return the datetime truncated to the day. Otherwise, the original
+     * value will be returned.
+     */
+    TRUNCATE_TEMPORAL_TO_DAY("DAY", new DateTimeValueFormatter("yyyy-MM-dd")),
+    
+    /**
+     * A {@link UniqueGranularity} implementation that, if provided a datetime value, will return the datetime truncated to the hour. Otherwise, the original
+     * value will be returned.
+     */
+    TRUNCATE_TEMPORAL_TO_HOUR("HOUR", new DateTimeValueFormatter("yyyy-MM-dd'T'HH")),
+    
+    /**
+     * A {@link UniqueGranularity} implementation that, if provided a datetime value, will return the datetime truncated to the minute. Otherwise, the original
+     * value will be returned.
+     */
+    TRUNCATE_TEMPORAL_TO_MINUTE("MINUTE", new DateTimeValueFormatter("yyyy-MM-dd'T'HH:mm"));
+    
+    private final String name;
+    private final Function<String,String> function;
```

Comment:
I love this.  This makes it easy to add additional enumerations down the road.

Reply:
Awesome, glad you like it.